protocols/upnp/src/behaviour.rs

```diff
@@ -0,0 +1,515 @@
+where
+    P: Provider,
+{
+    /// Gateway config.
+    config: Config,
+    /// UPnP interface state.
+    state: GatewayState<P>,
+
+    /// List of port mappings.
+    mappings: HashMap<Mapping, MappingState>,
+
+    /// Pending gateway events.
+    pending_events: FuturesUnordered<BoxFuture<'static, Event>>,
+}
+
+impl<P> Default for Behaviour<P>
+where
+    P: Provider + 'static,
+{
+    fn default() -> Self {
+        Self::new(Config::default())
+    }
+}
+
+impl<P> Behaviour<P>
+where
+    P: Provider + 'static,
+{
+    /// Builds a new `UPnP` behaviour.
+    pub fn new(config: Config) -> Self {
+        Self {
+            config,
+            state: GatewayState::Searching(P::search_gateway(config).boxed()),
+            mappings: Default::default(),
+            pending_events: Default::default(),
+        }
+    }
+}
+
+impl<P> NetworkBehaviour for Behaviour<P>
+where
+    P: Provider + 'static,
+{
+    type ConnectionHandler = dummy::ConnectionHandler;
+
+    type ToSwarm = ();
+
+    fn handle_established_inbound_connection(
+        &mut self,
+        _connection_id: ConnectionId,
+        _peer: PeerId,
+        _local_addr: &Multiaddr,
+        _remote_addr: &Multiaddr,
+    ) -> Result<libp2p_swarm::THandler<Self>, ConnectionDenied> {
+        Ok(dummy::ConnectionHandler)
+    }
+
+    fn handle_established_outbound_connection(
+        &mut self,
+        _connection_id: ConnectionId,
+        _peer: PeerId,
+        _addr: &Multiaddr,
+        _role_override: Endpoint,
+    ) -> Result<libp2p_swarm::THandler<Self>, libp2p_swarm::ConnectionDenied> {
+        Ok(dummy::ConnectionHandler)
+    }
+
+    fn on_swarm_event(&mut self, event: FromSwarm<Self::ConnectionHandler>) {
+        match event {
+            FromSwarm::NewListenAddr(NewListenAddr {
+                listener_id,
+                addr: multiaddr,
+            }) => {
+                let (addr, protocol) = match multiaddr_to_socketaddr_protocol(multiaddr.clone()) {
+                    Ok(addr_port) => addr_port,
+                    Err(_) => {
+                        log::debug!("multiaddress not supported for UPnP {multiaddr}");
+                        return;
+                    }
+                };
+
+                if let Some((mapping, _state)) = self
+                    .mappings
+                    .iter()
+                    .find(|(mapping, _state)| mapping.internal_addr.port() == addr.port())
+                {
+                    log::debug!("port from multiaddress {multiaddr} is already being mapped to another multiaddr: {}", mapping.multiaddr);
+                    return;
+                }
+
+                match &self.state {
+                    GatewayState::Searching(_) => {
+                        // As the gateway is not yet available we add the mapping with `MappingState::Inactive`
+                        // so that when and if it becomes available we map it.
+                        self.mappings.insert(
```

Comment:
If you add a mappings field to `Searching`, you could imply that they are all inactive and transition them to a new state once the gateway state changes.

Reply:
> yeah, but we'd still need `Inactive` for when a remapping fails and we want to try again later. Do you still feel we should do that?

Would it make sense to label these as failed mappings instead?